dask/array/blockwise.py

```diff
@@ -1,12 +1,74 @@
 import numbers
+import itertools
 import warnings
 
 import tlz as toolz
 
 from .. import base, utils
 from ..delayed import unpack_collections
 from ..highlevelgraph import HighLevelGraph
-from ..blockwise import blockwise as core_blockwise
+from ..blockwise import (
+    Blockwise,
+    blockwise as core_blockwise,
+    blockwise_token,
+)
+
+
+class BlockwiseCreateArray(Blockwise):
+    """
+    Specialized Blockwise Layer for array creation routines.
+
+    Enables HighLevelGraph optimizations.
+
+    Parameters
+    ----------
+    name: string
+        The output name.
+    func : callable
+        Function to apply to populate individual blocks. This function should take
+        an iterable containing the dimensions of the given block.
+    shape: iterable
+        Iterable containing the overall shape of the array.
+    chunks: iterable
+        Iterable containing the chunk sizes along each dimension of array.
+    """
+
+    def __init__(
+        self,
+        name,
+        func,
+        shape,
+        chunks,
+    ):
+        self.name = name
+        io_name = "blockwise-create-" + name
+
+        # Define "blockwise" graph
+        dsk = {self.name: (func, blockwise_token(0))}
+
+        # Define mapping between keys and chunks
+        chunk_key_map = {}
+        for idx in itertools.product(*[range(len(c)) for c in chunks]):
+            ndim = len(chunks)
+            chunk = tuple(chunks[i][idx[i]] for i in range(ndim))
+            chunk_key_map[tuple(idx)] = chunk
```

Comment:
Sorry for the very slow look at this @rjzamora! One thing I'm struggling to understand here: the previous implementation of `BlockwiseCreateArray` produced a mapping between chunk keys and their shapes lazily, allowing a much smaller footprint for large arrays. If I understand this correctly, this now goes back to keeping all of the keys and chunk shapes around.

Now, I could envision this (and similar `Layer`s) packing/unpacking the data needed to create this keymap so that it can be done lazily. Is that where you see this going longer-term?

Reply:
Yeah - It seems that I was going out of the way to avoid passing `chunks` to every task (required for lazy `chunk` resolution), but I don't think that priority makes any sense.  We should be able to embed `chunks` in a simple function wrapper, and then use a similar logic as before.

Reply:
Yeah, that makes sense to me, `chunks` takes up a lot less space than the fully expanded mapping.

Going further -- this implementation only works for uniform functions (ones, zeros, etc), but I have some work-in-progress with the old `BlockwiseIO` setup that takes into account the position of each chunk. In that case that data will also need to be constructed on the fly. I don't see it as a major impediment with the way you have set up things here, though.

Reply:
One more thought: we should add some basic sanity tests ensuring that all the layer implementations in `dask/dask` can be serialized so we don't get into trouble again :)

I'd be happy to do that in a follow-up, or make a PR to your branch here.

Reply:
> chunks takes up a lot less space than the fully expanded mapping.

It is worth considering that the expanded mapping does not need to be duplicated in every task when the graph is materialized, but `chunks` does need to be duplicated. Therefore, we may actually have an even larger memory footprint with `chunk_key_map` removed.  

> One more thought: we should add some basic sanity tests ensuring that all the layer implementations in dask/dask can be serialized so we don't get into trouble again :)

Yeah - It's actually a bit silly that this is not the first thing I did in this PR :)

Reply:
> **EDIT: Yeah - Embedding `chunks` in every task was not a scalable solution.

Interesting. I had been thinking about reducing the networking overhead in communicating between the client and the scheduler. Is it not possible to do the expansion on the scheduler and only produce each set of args on the fly?

Reply:
>Interesting. I had been thinking about reducing the networking overhead in communicating between the client and the scheduler. Is it not possible to do the expansion on the scheduler and only produce each set of args on the fly?

Sorry - I was just saying that we couldn't embed `chunks` in a callable object that is included in every task (so, I needed to revert that last commit).  However, I do agree that we can do better than the current `chunk_key_map` approach.  I am thinking we can allow the `io_deps` element (for a collection) to contain an element like `"callable".  If this `"callable"` element exists, the scheduler will generate the arguments for each partition by passing the index and whatever elements are stored in `io_deps` for that index (if any), to the callable function.... I will push a commit with this change, but feel free to suggest other solutions :)

Reply:
Although this "func" approach seems to work, it is only a minor optimization.  It will also require a bit more work to add a layer of protection to ensure that the user cannot relpace the callable object with arbitrary python code (a security vulnerability).  Therefore, I suggest we attack this in a separate/follow-up PR

Reply:
I think we have been dancing around just implementing some extra logic in the `__dask_distributed_{pack, unpack}__` for `BlockwiseCreateArray`. If we store the chunks at the top level, pack and unpack them simply, then produce the key map as needed, we could satisfy your constraint to not duplicate the chunks on every task, as well as mine to not send a mostly-materialized layer across the wire.

If we did that, we could also implement whatever gating/checks would be necessary to make sure only allowed "`func`"s get expanded on the scheduler. Or am I misunderstanding what you mean @rjzamora?

Reply:
Thanks @ian-r-rose - I don't think you are misunderstanding.  `__dask_distributed_{pack, unpack}__` is certainly the most promising place to handle dynamic chunk generation.  The mechanism itself is very simple, but the gating/check is what I am not quite sure about yet (and have not prototyped anything yet).

Note that I opened #7281 with the intention to hash out these blockwise/array basics before adding DataFrame to the mix.  At first, I was thinking that we could start with the "mostly-materialized" layer approach, but I can probably revise that PR with a new proposal for dynamic chunk materialization tomorrow.

Reply:
Thanks @rjzamora, that makes sense to me, I may have some bandwidth this week to look at dynamic chunk materialization if that would be helpful

Reply:
Cool - I'm definitely interested to get your thoughts/ideas.

The main challenge is that a `BlockwiseCreateArray` will often become a vanilla `Blockwise` Layer after `optimize_blockwise` is called, so we cannot simply rely on code in `BlockwiseCreateArray.__dask_distributed_{pack, unpack}__`.    To understand why, consider the case where two arrays are generated with different creation techniques (with one possibly requiring non-trivial `io_deps`).  If the two arrays are combined with a blockwise addition operation, `optimize_blockwise` will fuse the three layers together, and the final layer type cannot always be `BlockwiseCreateArray`.

Since we are already planning to support the `io_deps` attribute for all `Blockwise` layers, it would be nice to use the same structure to specify dynamic "unpacking" options.  For example, we may be able to register a list of options (e.g `"generate_chunk"`) that will map onto "cleared" functions known by the scheduler:

```python
# Dynamic `io_deps` generation function
def generate_chunk(idx, chunks):
    ndim = len(chunks)
    return tuple(chunks[i][idx[i]] for i in range(ndim))

...

# "registered" function dict
io_dep_functions = {
    "generate_chunk" : generate_chunk,
    ...
}

# Example `io_deps` definition.
# Scheduler would know to call `generate_chunk(idx, ((4, 4, 2), (4, 4, 2)))`
# to construct the input for each chunk/index of 'ones-4627d3bbed71139f53cd1f686350e412'
io_deps = {'ones-4627d3bbed71139f53cd1f686350e412': {"generate_chunk": ((4, 4, 2), (4, 4, 2))}}
```